frontend/rust-lib/flowy-ai/src/ai_manager.rs

```diff
@@ -238,30 +244,161 @@ impl AIManager {
     let question_message_id = chat
       .get_question_id_from_answer_id(answer_message_id)
       .await?;
+
+    let preferred_model = self
+      .store_preferences
+      .get_object::<AIModel>(&ai_available_models_key(chat_id));
     chat
-      .stream_regenerate_response(question_message_id, answer_stream_port, format)
+      .stream_regenerate_response(
+        question_message_id,
+        answer_stream_port,
+        format,
+        preferred_model,
+      )
       .await?;
     Ok(())
   }
 
-  pub async fn get_available_models(&self) -> FlowyResult<ModelList> {
+  pub async fn get_workspace_select_model(&self) -> FlowyResult<String> {
     let workspace_id = self.user_service.workspace_id()?;
+    let model = self
+      .cloud_service_wm
+      .get_workspace_default_model(&workspace_id)
+      .await?;
+    Ok(model)
+  }
+
+  pub async fn get_server_available_models(&self) -> FlowyResult<Vec<String>> {
+    let workspace_id = self.user_service.workspace_id()?;
+
+    // First, try reading from the cache.
+    {
+      let cached_models = self.server_models.read().await;
+      if !cached_models.is_empty() {
+        return Ok(cached_models.clone());
+      }
+    }
+
+    // Cache miss: fetch from the cloud.
     let list = self
       .cloud_service_wm
       .get_available_models(&workspace_id)
       .await?;
-    Ok(list)
+    let models = list
+      .models
+      .into_iter()
+      .map(|m| m.name)
+      .collect::<Vec<String>>();
+
+    // Update the cache.
+    *self.server_models.write().await = models.clone();
+    Ok(models)
+  }
+
+  pub async fn update_selected_model(&self, source: String, model: AIModelPB) -> FlowyResult<()> {
+    let source_key = ai_available_models_key(&source);
+    self
+      .store_preferences
+      .set_object::<AIModel>(&source_key, &model.clone().into())?;
+
+    chat_notification_builder(&source, ChatNotification::DidUpdateSelectedModel)
+      .payload(model)
+      .send();
+    Ok(())
+  }
```

Comment:
For AI chats, should we save the selected model into the chat settings?

Reply:
I thought about it, but since not every device has local AI set up, I decided to simply store it locally.